Add `-datacarriernum` options

ariard opened the pull request.

ariard committed:
Add -datacarriernum

petertodd commented:
Two issues with this:

1. The `MAX_DATACARRIER_OUTPUTS = 100` limit is much too low for the spirit of libre relay. The default limit should be sufficiently high to be impossible to actually reach; put another way, it should be unlimited.
2. Every feature we add is a future maintenance burden. In my discussions with miners, what they want most from Libre Relay is good unit tests and a reasonable libre policy that is unlikely to cause them any issues. If they wanted a more limited OP_Return policy, they could just run core.

With that in mind, I'd suggest taking this pull-req and proposing it as a pull-req for Bitcoin Core.

ariard commented:
Pushed an update, addressing the 2 concerns.

> The MAX_DATACARRIER_OUTPUTS = 100 limit is much too low for the spirit of libre relay. The default limit should be > sufficiently high to be impossible to actually reach; put another way, it should be unlimited.

Added a new `MAX_DATACARRIER_OUTPUTS_HARDLIMIT=3_999_833` which means one minimal size transaction + 3_999_833 1-byte OP_RETURN. Current transaction issuers can at most make a transaction no bigger than `MAX_BLOCK_WEIGHT`. `MAX_DATACARRIER_OUTPUTS_HARDLIMIT` when we see network-level consensus upgrade in matters of block size.

> Every feature we add is a future maintenance burden. In my discussions with miners, what they want most from Libre Relay is good unit tests and a reasonable libre policy that is unlikely to cause them any issues. If they wanted a more limited OP_Return policy, they could just run core.

Added a new `MAX_DATACARRIER_OUTPUTS_SOFTLIMIT=1` which is matching the max number of OP_RETURN outputs accepted by Bitcoin Core as a transaction-relay policy since release v0.10.0. There is a new check in `src/node/mempool_args.cpp` verifying that node setting `datacarriernum` is no strictly superior than `MAX_DATACARRIER_OUTPTUS_HARDLIMIT`.

Distinction between hard and soft full-node resources limits for denial-of-service mitigation while allowing more flexibility for use-cases experimentations, including accepting transactions yielding higher-income block template is coming from my kernel hacking experience, e.g `getrlimit()`.

petertodd commented:
> Pushed an update, addressing the 2 concerns.

Thing is, you haven't really addressed the second concern. Every bit of code added to Libre Relay is code that needs to be cherry-picked and updated every time Bitcoin Core makes a release; I'm working on that for v27.0rc1 right now. I just don't see a good reason to make the # of OP_Return outputs configurable. What person would want to relay/mine more than 1, but less than ∞? Simply commenting out that check is much simpler, and less work to maintain.

ariard commented:
> What person would want to relay/mine more than 1, but less than ∞? Simply commenting out that check is much simpler, and less work to maintain.

Any person wishing to operate transaction-relay traffic shaping, where you would like to apply quantitative segments over global transaction-relay traffic (e.g 1/3 third lightning, 1/3 third datacarrier-using colored coins, 1/3 usual transactions). As each segment can have it’s own time-sensitive constraints / liquidity preferences and you wish to optimize income.

As of today, I’m not aware of any mining pools / block template constructors doing so, this is true it sounds very too much flexibility for now. If libre relay focus on maintenance, correct `-datacarrierenum` might be superfluous for now. Going further, why not removing max OP_RETURN check completely rather than commenting out ? Always easy to re-insert it in mempool’s `PreChecks()` code path.

cryptoquick commented on src/init.cpp:
```diff
@@ -593,6 +593,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-acceptstalefeeestimates", strprintf("Read fee estimates even if they are stale (%sdefault: %u) fee estimates are considered stale if they are %s hours old", "regtest only; ", DEFAULT_ACCEPT_STALE_FEE_ESTIMATES, Ticks<std::chrono::hours>(MAX_FILE_AGE)), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-bytespersigop", strprintf("Equivalent bytes per sigop in transactions for relay and mining (default: %u)", DEFAULT_BYTES_PER_SIGOP), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
     argsman.AddArg("-datacarrier", strprintf("Relay and mine data carrier transactions (default: %u)", DEFAULT_ACCEPT_DATACARRIER), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
+    argsman.AddArg("-datacarriernum", strprintf("Relay and mine transacrtions with max %u data carrier outputs", MAX_DATACARRIER_OUTPUTS_SOFTLIMIT), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
```
```suggestion
    argsman.AddArg("-datacarriernum", strprintf("Relay and mine transactions with max %u data carrier outputs", MAX_DATACARRIER_OUTPUTS_SOFTLIMIT), ArgsManager::ALLOW_ANY, OptionsCategory::NODE_RELAY);
```

petertodd commented:
> Going further, why not removing max OP_RETURN check completely rather than commenting out ?

IIRC I wanted it to be clear what was there for reference for future rebases. But I don't think I've followed that consistently anyway...

petertodd commented:
Closing. No plans to implement this.

ariard commented:
>IIRC I wanted it to be clear what was there for reference for future rebases. But I don't think I've followed that consistently anyway…

Yes, and as you said in the OP as every new feature is a maintenance burden, it's a valuable point not to consider `datacarrierenum` for now for inclusion in `libre-relay`. For what it’s worth, an example of use-case for which many data outputs (i.e `datacarrierenum`) could be worthwhile in the future: https://github.com/ordinals/ord/issues/2405

petertodd commented:
> > IIRC I wanted it to be clear what was there for reference for future rebases. But I don't think I've followed that consistently anyway…
> 
> Yes, and as you said in the OP as every new feature is a maintenance burden, it's a valuable point not to consider `datacarrierenum` for now for inclusion in `libre-relay`. For what it’s worth, an example of use-case for which many data outputs (i.e `datacarrierenum`) could be worthwhile in the future: [ordinals/ord#2405](https://github.com/ordinals/ord/issues/2405)

Libre Relay supports an unlimited number of `OP_Return` outputs out-of-the-box. The only thing a `datacarrierenum` option would do is add a limit, which is not the goal of Libre Relay.
